### ISSUE TITLE
Mitmproxy doesn't check revocation lists in 'verify' mode

### ISSUE THREAD

**Issue by ColinChartier** [mitmproxy/mitmproxy#2235]
To reproduce, simply use mitmproxy to access https://revoked.badssl.com/

**Comment by mhils** [comment]
Thanks for raising this. We internally just use OpenSSL to handle certificate verification. To the best of my knowledge, OpenSSL does not provide any means (beyond low-level crypto stuff) to check if a certificate has been revoked [[1]](https://stackoverflow.com/questions/2913440/openssl-how-to-check-if-a-certificate-is-revoked-or-not). If it were as simple as flipping a switch I'd be happy to do that immediately, but I don't feel comfortable  writing the custom revocation logic that handles all the tricky edge-cases correctly. Unfortunately, certificate revocation is still often fundamentally broken in practice [[2]](https://www.imperialviolet.org/2014/04/19/revchecking.html). 

Long story short, we currently do what (almost?) every non-browser software is doing: We don't check for certificate revocations. I agree it's not ideal, yet it's what curl and everything else that uses OpenSSL does. Contributions are welcome of course.

**Comment by ravish-bhatia** [comment]
Hi,

Is this issue still open?

**Comment by ravish-bhatia** [comment]
I have been working on this issue, and haven't succeeded much as of now. 
So, another approach which I am planning to implement involves following test case:
- get verisign(or any root CA) crl list/file on local machine 
- when mitmproxy acts as a client, it verify the certificate against this CRL file on the machine 
- This is just a test case, it is not easy to get all the root CA's crl file on the local machine. But it is just an idea, and if succeeded we might be able to pass through this issue to some extent

Need reviews on this before I proceed further

**Comment by mhils** [comment]
@ravish-bhatia, thanks for your interest in fixing this. This issue unfortunately is not an isolated mitmproxy issue, see for example Scott Helme's fantastic essay on why [revocation is broken](https://scotthelme.co.uk/revocation-is-broken/). I strongly believe that this issue fundamentally needs to be fixed outside of mitmproxy before we can make any progress here.

Coming back to your proposal, let me point out a few issues:
- How do we get CRLs on the local machine, how are they updated? What happens if updates fail?
- From a privacy perspective, mitmproxy deliberatly does not phone home at the moment. We probably do not want to phone home to CAs on startup either.
- CRLs are pretty much deprecated: https://groups.google.com/forum/#!msg/mozilla.dev.security.policy/85MV81ch2Zo/fmqckeHESKUJ
- etc.

I really appreciate that you are trying to tackle a difficult problem here, but I would recommend that you either shift your attention to mitigating this in OpenSSL (or some other TLS library) itself or, if you are interested in contributing to mitmproxy specifically, look for different issues. For example, our web ui could really use a certificate display component. 😉

**Comment by ravish-bhatia** [comment]
@mhils Thank you for the elaborated reply, I understand your point. However, I could not get the second point, why would phone home to CAs would be a privacy concern?

**Comment by mhils** [comment]
@ravish-bhatia: It's not a massive issue, but given how easily fingerprintable HTTP traffic is, we would give CAs an idea of who on the internet uses mitmproxy. We also have researchers that work in sensitive environments where a whole bunch of CRL requests would give away that they are using mitmproxy. This is not a dealbreaker in my eyes, but something we need to consider. I'd personally love to have limited telemetry in mitmproxy to help us with development, but I know that for example @cortesi cares deeply about this and holds a strong "no" stance.
